### PR TITLE
feat: better handling of cost property calculation corner cases

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR enables the Azure property provider to raise warnings/errors as conditions when handling cost calculation corner cases.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Unit tests
- [x] Integration tests

### CODE REVIEW

**Comment by michaelawyu** [comment]
**Important**: there are a few additional PRs that will be sent separately to enable E2E and upgrade tests for functionalities in this PR.

Also -> this PR involves a condition type change; will send a separate PR to set up the cleanup process (with upgrade tests for verification).

File: pkg/propertyprovider/azure/provider_integration_test.go [hunk 1]
```diff
@@ -172,29 +172,88 @@ var (
...
-					Properties: map[clusterv1beta1.PropertyName]clusterv1beta1.PropertyValue{
-						propertyprovider.NodeCountProperty: {
-							Value: fmt.Sprintf("%d", len(nodes)),
+				perCPUCoreCost := "0.0"
+				perGBMemoryCost := "0.0"
+				costCollectionWarnings := []string{}
+				var costCollectionErr error
+
+				switch {
+				case len(nodes) == 0:
```
Comment (britaniar) [inline pkg/propertyprovider/azure/provider_integration_test.go:209]: So we do nothing in this case?

Comment (britaniar) [reply pkg/propertyprovider/azure/provider_integration_test.go:209]: Do we want a log at least?

File: pkg/propertyprovider/azure/provider_integration_test.go [hunk 2]
```diff
@@ -172,29 +172,88 @@ var (
...
-						return fmt.Errorf("on-demand price not found for SKU %s", node.Labels[trackers.AKSClusterNodeSKULabelName])
+						missingSKUSet[sku] = true
 					}
 					totalCost += cost
 				}
-				perCPUCost := fmt.Sprintf(CostPrecisionTemplate, totalCost/totalCPUCores)
-				perGBMemoryCost := fmt.Sprintf(CostPrecisionTemplate, totalCost/totalMemoryGBs)
+				missingSKUs := []string{}
+				for sku := range missingSKUSet {
+					missingSKUs = append(missingSKUs, sku)
```
Comment (britaniar) [inline pkg/propertyprovider/azure/provider_integration_test.go:199]: Seems like the integration test is failing around here. May need to sort so that the list is consistent.

File: pkg/propertyprovider/azure/trackers/nodes.go [hunk 3]
```diff
@@ -129,10 +133,13 @@ func (nt *NodeTracker) calculateCosts() {
 
 	// Sum up the total costs.
 	totalHourlyRate := 0.0
+	missingSKUs := make([]string, 0)
 	for sku, ns := range nt.nodeSetBySKU {
 		hourlyRate, found := nt.pricingProvider.OnDemandPrice(sku)
```
Comment (ryanzhang-oss) [inline pkg/propertyprovider/azure/trackers/nodes.go:162]: the new lib will return a fake value instead of reporting notFound

File: pkg/propertyprovider/azure/provider_integration_test.go (hunk 1 above)
Comment (michaelawyu) [reply pkg/propertyprovider/azure/provider_integration_test.go:209]: Yeah, it is considered normal (no nodes, no node related costs); we do have logs in the provider node controller to confirm on the no nodes situation, but adding a log here should help too -> will make the edit

File: pkg/propertyprovider/azure/provider_integration_test.go (hunk 2 above)
Comment (michaelawyu) [reply pkg/propertyprovider/azure/provider_integration_test.go:199]: Yeah, my bad. I forgot that the tests now are comparing the output string verbatim -> will add the sort.

File: pkg/propertyprovider/azure/trackers/nodes.go (hunk 3 above)
Comment (michaelawyu) [reply pkg/propertyprovider/azure/trackers/nodes.go:162]: Hi Ryan!

Their code,

```
// OnDemandPrice returns the last known on-demand price for a given instance type, returning false if there is no
// known on-demand pricing for the instance type.
func (p *Provider) OnDemandPrice(instanceType string) (float64, bool) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	price, ok := p.onDemandPrices[instanceType]
	if !ok {
		// if we don't have a price, check if it's a known SKU with missing price
		if price, ok = skusWithMissingPrice[instanceType]; ok {
			return price, true
		}
		return 0.0, false
	}
	return price, true
}
```

only covers the known missing entries, so we still need to count for unknown cases our side.

Comment (michaelawyu) [reply pkg/propertyprovider/azure/trackers/nodes.go:162]: Though like you mentioned, Karpenter returning 999 would make it look quite weird on our side. We haven't upgraded to that version as I understand, but...

File: pkg/propertyprovider/azure/provider_integration_test.go (hunk 1 above)
Comment (michaelawyu) [reply pkg/propertyprovider/azure/provider_integration_test.go:209]: A log entry has been added to the `calculateCosts()` function. (The current code selection is part of the integration test).

File: pkg/propertyprovider/azure/trackers/nodes.go (hunk 3 above)
Comment (michaelawyu) [reply pkg/propertyprovider/azure/trackers/nodes.go:162]: The code has been updated to reflect the changes in the latest Karpenter/Az provider version.

File: pkg/propertyprovider/azure/provider.go [hunk 4]
```diff
@@ -203,17 +202,20 @@ func (p *PropertyProvider) Collect(_ context.Context) propertyprovider.PropertyC
...
-		// track the previously returned condition. A timestamp will be added in the upper layer.
+	perCPUCost, perGBMemoryCost, warnings, err := p.nodeTracker.Costs()
+	switch {
+	case err != nil:
+		// An error occurred when calculating costs; do no set the cost properties and
+		// track the error.
 		conds = append(conds, metav1.Condition{
-			Type:    PropertyCollectionSucceededConditionType,
+			Type:    CostPropertiesCollectionSucceededCondType,
 			Status:  metav1.ConditionFalse,
```
Comment (ryanzhang-oss) [inline pkg/propertyprovider/azure/provider.go:212]: what happens when this false?

File: pkg/propertyprovider/azure/provider.go [hunk 5]
```diff
@@ -222,6 +224,29 @@ func (p *PropertyProvider) Collect(_ context.Context) propertyprovider.PropertyC
 			Value:           fmt.Sprintf(CostPrecisionTemplate, perGBMemoryCost),
 			ObservationTime: metav1.Now(),
 		}
+		conds = append(conds, metav1.Condition{
+			Type:    CostPropertiesCollectionSucceededCondType,
+			Status:  metav1.ConditionTrue,
+			Reason:  CostPropertiesCollectionDegradedReason,
```
Comment (ryanzhang-oss) [inline pkg/propertyprovider/azure/provider.go:230]: since the property selector doesn't read this reason so it will take this value as IS? How does the user know?

File: pkg/propertyprovider/azure/provider.go (hunk 4 above)
Comment (michaelawyu) [reply pkg/propertyprovider/azure/provider.go:212]: Hi Ryan! When this is set to False, users will see the condition with the error details and no cost data will be exposed for the cluster (the properties themselves will not be added) -> if users have cost property based filters (e.g., pick clusters with cpu prices <1), such clusters will be ignored.

File: pkg/propertyprovider/azure/provider.go (hunk 5 above)
Comment (michaelawyu) [reply pkg/propertyprovider/azure/provider.go:230]: Hi Ryan! Yeah, this is consistent with the behavior before, where we would take the data returned from Karpenter as they are (even if they are stale), and if the cluster has some SKUs that cannot be recognized, we consider them to be free of cost.

Comment (michaelawyu) [reply pkg/propertyprovider/azure/provider.go:230]: I totally understand the concern here of course; later we might be able to offer a way to filter out clusters with degraded properties... we do not support boolean typed properties yet, 0/1-based properties can do, but they might not look very good.

**Comment by michaelawyu** [comment]
Note: sync'd offline with ryanzhangoss; the behavior has been slightly tweaked: considering that the system works best when handling homogeneous clusters (clusters using nodes of exactly one SKU), the property provider will now report partially missing SKUs as an error (i.e., no costs will be reported; the scheduler will ignore such clusters when processing cost-related queries) rather than returning skewed cost data.

**Review by ryanzhang-oss — APPROVED** [review]

File: pkg/propertyprovider/azure/provider.go [hunk 6]
```diff
@@ -55,14 +55,13 @@ const (
...
 	// The condition related values in use by the Azure property provider.
-
-	// PropertyCollectionSucceededConditionType is a condition type that indicates whether a
-	// property collection attempt has succeeded.
-	PropertyCollectionSucceededConditionType         = "AKSClusterPropertyCollectionSucceeded"
-	PropertyCollectionSucceededReason                = "AllPropertiesCollectedSuccessfully"
-	PropertyCollectionFailedCostErrorReason          = "FailedToCollectCosts"
-	PropertyCollectionSucceededMessage               = "All properties have been collected successfully"
-	PropertyCollectionFailedCostErrorMessageTemplate = "An error has occurred when collecting cost properties: %v"
+	CostPropertiesCollectionSucceededCondType   = "AKSClusterCostPropertiesCollectionSucceeded"
```
Comment (ryanzhang-oss) [inline pkg/propertyprovider/azure/provider.go:58]: I didn't notice the condition name is changed...
